### PR TITLE
Added support for file-based variant template files.

### LINKED ISSUE #2: Support file based variants.
Problem
* Fractal supports [file-based variant files](http://fractal.build/guide/components/variants#creating-file-based-variants), but they are not supported by twig_fractal right now.

Goal
* Ensure feature parity with Fractal, so that users are not limited or confused.

Notes
* Variants currently can only be specified with `componentname--variant.twig`

### PR DESCRIPTION
**Pull request by fabianmarz**
fixes #2

### CODE REVIEW

File: src/Component.php [hunk 1]
```diff
@@ -128,6 +129,8 @@ protected function loadDefinition(string $pathname) {
   /**
    * Returns the relative file path of the Fractal YAML configuration file for a given component name.
    *
+   * While Fractal do not support file-based variant component configuration files, neither we do.
```
Comment (sun) [inline src/Component.php]: If both sides don't do anything, then the language construct of _"While …"_ does not make sense.  It only works if one side is the opposite, as in _"While xyz does something, we don't."_

```
Neither Fractal nor this implementation supports separate configuration files per variant.
```

File: src/Component.php [hunk 2]
```diff
@@ -178,23 +181,27 @@ protected function getVariantDefaultVariables(string $modifier, array $defined_v
...
+   * Returns the component's compound name, name, and a list of variants.
    *
    * @param string $compound_name
    *   A compound name including the component and optionally variants, delimited
    *   by double-hyphens (`--`).
    *
    * @return array
    *   An array with three elements:
-   *   1. the component's pathname
+   *   1. the component compound name
```
Comment (sun) [inline src/Component.php]: It does actually return the pathname which is then assigned to the property `$this->pathname` – why did you change this?

File: src/Component.php (hunk 1 above)
Comment (fabianmarz) [reply src/Component.php]: Ouch, I first wanted to use "since" but decided to go with "while" 🙈. Thanks for pointing out. I wanted to be clear why this implementation do not support separate config files. So I'd rather exchange the while with a since. Or would you rather go with your suggestion?

File: src/Component.php (hunk 2 above)
Comment (fabianmarz) [reply src/Component.php]: Yes and no. The compound name consists of e.g. `@molecules/teaser/teaser--compact.twig` (but should maybe just `teaser--compact.twig`). So if we consider this full string as the path, we should rename the parameter name as well, and get rid of the confusing `$compound_name` variable name. I'd rather got with `$pathname`, thanks for pointing out 👌.

File: src/Component.php (hunk 1 above)
Comment (sun) [reply src/Component.php]: If both don't do / do something, it's too much "because".  Go with the "neither" statement instead, or simplify furher:

```diff
-Neither Fractal nor this implementation supports separate configuration files per variant.
+Like Fractal, this implementation does not support separate configuration files per variant.
```

File: src/Component.php (hunk 2 above)
Comment (sun) [reply src/Component.php]: The passed argument is the originally passed pathname (user input) from the calling code / twig template.  It might be a virtual name containing one or more variant suffixes.

The returned pathname is the validated canonical pathname of the component.  It is assigned to the property and later used for reading the component's definition/configuration file.

File: src/Component.php [hunk 3]
```diff
@@ -193,13 +193,16 @@ protected function getVariantDefaultVariables(string $modifier, array $defined_v
...
-  protected function extractParts(string $pathname): array {
+  protected function extractParts(string $compound_name): array {
     $library = Drupal::service('twig.loader.componentlibrary');
-    $variants = explode('--', basename(basename($pathname, '.twig'), '.html'));
+    $variants = explode('--', basename(basename($compound_name, '.twig'), '.html'));
     $component = array_shift($variants);
     // If no file-based variant template file exist use the base component otherwise.
-    if (!$library->exists($pathname)) {
-      $pathname = preg_replace('@--[^.]+@', '', $pathname);
+    if (!$library->exists($compound_name)) {
```
Comment (fabianmarz) [inline src/Component.php]: For readability, I decided to go with a full if/else statement over a ternary. Also I wasn't very happy with declaring the `$pathname` variable before and have just a if condition instead.

File: src/Component.php [hunk 4]
```diff
@@ -129,7 +129,7 @@ protected function loadDefinition(string $pathname) {
   /**
    * Returns the relative file path of the Fractal YAML configuration file for a given component name.
    *
-   * While Fractal do not support file-based variant component configuration files, neither we do.
+   * Like Fractal, this implementation does not support separate configuration files per variant.
```
Comment (sun) [inline src/Component.php]: Still exceeds 80 chars

File: src/Component.php [hunk 5]
```diff
@@ -191,9 +194,16 @@ protected function getVariantDefaultVariables(string $modifier, array $defined_v
...
-    $pathname = preg_replace('@--[^.]+@', '', $compound_name);
+    $library = Drupal::service('twig.loader.componentlibrary');
     $variants = explode('--', basename(basename($compound_name, '.twig'), '.html'));
     $component = array_shift($variants);
+    // If no file-based variant template file exist use the base component otherwise.
+    if (!$library->exists($compound_name)) {
+      $pathname = preg_replace('@--[^.]+@', '', $compound_name);
+    }
+    else {
+      $pathname = $compound_name;
```
Comment (sun) [inline src/Component.php]: As the returned pathname is used for loading the component's definition/configuration file, and as variant-specific definition/configuration files are not supported (as documented), I don't understand why we would ever want to return the compound name (including variants) as pathname?

Comment (fabianmarz) [reply src/Component.php]: As the PR title and the related issue states, we want to support file based variant files, but no dedicated configuration files. As you can see [here](https://github.com/netzstrategen/drupal-twig_fractal/blob/master/src/Node/Render.php#L68) the node is calling `$component->getPathname()` as the template value.

Comment (sun) [reply src/Component.php]: As you can see [here](https://github.com/netzstrategen/drupal-twig_fractal/blob/master/src/Component.php#L60) the Component is calling `$component->getPathname()` to load the config file.

Looks like property overloading.  We probably need a separate method `getTemplatePathname()` then.

File: src/Component.php (hunk 4 above)
Comment (fabianmarz) [reply src/Component.php]: Is this a deal breaker? The comment 2 lines prior also exceeds the 80 character limit. I can add line breaks if wanted. 😅

File: src/Component.php (hunk 5 above)
Comment (fabianmarz) [reply src/Component.php]: I also changed the config loading code in this PR:

```diff
- return $path['dirname'] . '/' . $path['filename'] . '.config.yml';
+ return $path['dirname'] . '/' . $this->getName() . '.config.yml';
```

Changing the argument seems to be a more valid solution to me. 👍

File: src/Component.php [hunk 6]
```diff
@@ -121,12 +129,18 @@ protected function mergeContext(array $base, array $override): array {
    *   The parsed component definition.
    */
   protected function loadDefinition(string $pathname) {
+    $component_data = [];
     $definition_pathname = $this->getDefinitionFilePath($pathname);
-    return Yaml::parse(file_get_contents($definition_pathname));
+    if (file_exists($definition_pathname)) {
```
Comment (fabianmarz) [inline src/Component.php:134]: There was an issue when the config file did not exist an error was thrown. So I added an additional `file_exist` check.

File: src/Component.php (hunk 4 above)
Comment (sun) [reply src/Component.php]: The first line of a PHPDoc block is called the PHPDoc summary line.

The PHPDoc summary line must not wrap.

File: src/Component.php [hunk 7]
```diff
@@ -122,12 +129,16 @@ protected function mergeContext(array $base, array $override): array {
...
+      $component_data = Yaml::parse(file_get_contents($definition_pathname));
+    }
+    return $component_data;
   }
 
   /**
-   * Returns the relative file path of the Fractal YAML configuration file for a given component name.
+   * Returns the relative file path of the Fractal YAML configuration file fora given component name.
```
Comment (sun) [inline src/Component.php]: Unintended change here

File: src/Component.php (hunk 6 above)
Comment (sun) [reply src/Component.php:134]: What's the behavior if the config file doesn't exist?  Is the error reported?

Comment (sun) [reply src/Component.php:134]: Is it an error to begin with?

File: src/Component.php [hunk 8]
```diff
@@ -187,14 +201,19 @@ protected function getVariantDefaultVariables(string $modifier, array $defined_v
...
    */
   protected function extractParts(string $compound_name): array {
-    $pathname = preg_replace('@--[^.]+@', '', $compound_name);
+    $library = Drupal::service('twig.loader.componentlibrary');
     $variants = explode('--', basename(basename($compound_name, '.twig'), '.html'));
     $component = array_shift($variants);
-    return [$pathname, $component, $variants];
+    $pathname = $templatePathname = preg_replace('@--[^.]+@', '', $compound_name);
+    if ($library->exists($compound_name)) {
+      $templatePathname = $compound_name;
```
Comment (sun) [inline src/Component.php]: This would be a bit more readable & clear:

```php
    $pathname = preg_replace('@--[^.]+@', '', $compound_name);
    $templatePathname = $library->exists($compound_name) ? $compound_name : $pathname;
```
I'd also move the two assignments back to the beginning of the method to match the order of return values and docs.

File: src/Component.php (hunk 6 above)
Comment (fabianmarz) [reply src/Component.php:134]: No it's not reported. Also I don't see necessity as some components don't have a config file by purpose.

**Review by sun — APPROVED** [review]
Added b9e1db9c976e91eb00e5da8043362843af548a18 to fix remaining issues.
